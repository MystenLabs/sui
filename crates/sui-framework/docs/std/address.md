---
title: Module `std::address`
---

Provides a way to get address length since it's a
platform-specific parameter.


-  [Function `length`](#std_address_length)


<pre><code></code></pre>



<a name="std_address_length"></a>

## Function `length`

Should be converted to a native function.
Current implementation only works for Sui.


<pre><code><b>public</b> <b>fun</b> <a href="../std/address.md#std_address_length">length</a>(): <a href="../std/u64.md#std_u64">u64</a>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/address.md#std_address_length">length</a>(): <a href="../std/u64.md#std_u64">u64</a> {
    32
}
</code></pre>



</details>
