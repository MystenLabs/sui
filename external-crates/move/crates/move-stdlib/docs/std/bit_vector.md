
<a name="std_bit_vector"></a>

# Module `std::bit_vector`



-  [Struct `BitVector`](#std_bit_vector_BitVector)
-  [Constants](#@Constants_0)
-  [Function `new`](#std_bit_vector_new)
-  [Function `set`](#std_bit_vector_set)
-  [Function `unset`](#std_bit_vector_unset)
-  [Function `shift_left`](#std_bit_vector_shift_left)
-  [Function `is_index_set`](#std_bit_vector_is_index_set)
-  [Function `length`](#std_bit_vector_length)
-  [Function `longest_set_sequence_starting_at`](#std_bit_vector_longest_set_sequence_starting_at)


<pre><code></code></pre>



<a name="std_bit_vector_BitVector"></a>

## Struct `BitVector`



<pre><code><b>public</b> <b>struct</b> <a href="../std/bit_vector.md#std_bit_vector_BitVector">BitVector</a> <b>has</b> <b>copy</b>, drop, store
</code></pre>



<details>
<summary>Fields</summary>


<dl>
<dt>
<code><a href="../std/bit_vector.md#std_bit_vector_length">length</a>: <a href="../std/u64.md#std_u64">u64</a></code>
</dt>
<dd>
</dd>
<dt>
<code>bit_field: <a href="../std/vector.md#std_vector">vector</a>&lt;bool&gt;</code>
</dt>
<dd>
</dd>
</dl>


</details>

<a name="@Constants_0"></a>

## Constants


<a name="std_bit_vector_EINDEX"></a>

The provided index is out of bounds


<pre><code><b>const</b> <a href="../std/bit_vector.md#std_bit_vector_EINDEX">EINDEX</a>: <a href="../std/u64.md#std_u64">u64</a> = 131072;
</code></pre>



<a name="std_bit_vector_ELENGTH"></a>

An invalid length of bitvector was given


<pre><code><b>const</b> <a href="../std/bit_vector.md#std_bit_vector_ELENGTH">ELENGTH</a>: <a href="../std/u64.md#std_u64">u64</a> = 131073;
</code></pre>



<a name="std_bit_vector_MAX_SIZE"></a>

The maximum allowed bitvector size


<pre><code><b>const</b> <a href="../std/bit_vector.md#std_bit_vector_MAX_SIZE">MAX_SIZE</a>: <a href="../std/u64.md#std_u64">u64</a> = 1024;
</code></pre>



<a name="std_bit_vector_WORD_SIZE"></a>



<pre><code><b>const</b> <a href="../std/bit_vector.md#std_bit_vector_WORD_SIZE">WORD_SIZE</a>: <a href="../std/u64.md#std_u64">u64</a> = 1;
</code></pre>



<a name="std_bit_vector_new"></a>

## Function `new`



<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_new">new</a>(<a href="../std/bit_vector.md#std_bit_vector_length">length</a>: <a href="../std/u64.md#std_u64">u64</a>): <a href="../std/bit_vector.md#std_bit_vector_BitVector">std::bit_vector::BitVector</a>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_new">new</a>(<a href="../std/bit_vector.md#std_bit_vector_length">length</a>: <a href="../std/u64.md#std_u64">u64</a>): <a href="../std/bit_vector.md#std_bit_vector_BitVector">BitVector</a> {
    <b>assert</b>!(<a href="../std/bit_vector.md#std_bit_vector_length">length</a> &gt; 0, <a href="../std/bit_vector.md#std_bit_vector_ELENGTH">ELENGTH</a>);
    <b>assert</b>!(<a href="../std/bit_vector.md#std_bit_vector_length">length</a> &lt; <a href="../std/bit_vector.md#std_bit_vector_MAX_SIZE">MAX_SIZE</a>, <a href="../std/bit_vector.md#std_bit_vector_ELENGTH">ELENGTH</a>);
    <b>let</b> <b>mut</b> counter = 0;
    <b>let</b> <b>mut</b> bit_field = <a href="../std/vector.md#std_vector_empty">vector::empty</a>();
    <b>while</b> (counter &lt; <a href="../std/bit_vector.md#std_bit_vector_length">length</a>) {
        bit_field.push_back(<b>false</b>);
        counter = counter + 1;
    };
    <a href="../std/bit_vector.md#std_bit_vector_BitVector">BitVector</a> {
        <a href="../std/bit_vector.md#std_bit_vector_length">length</a>,
        bit_field,
    }
}
</code></pre>



</details>

<a name="std_bit_vector_set"></a>

## Function `set`

Set the bit at <code>bit_index</code> in the <code>bitvector</code> regardless of its previous state.


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_set">set</a>(bitvector: &<b>mut</b> <a href="../std/bit_vector.md#std_bit_vector_BitVector">std::bit_vector::BitVector</a>, bit_index: <a href="../std/u64.md#std_u64">u64</a>)
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_set">set</a>(bitvector: &<b>mut</b> <a href="../std/bit_vector.md#std_bit_vector_BitVector">BitVector</a>, bit_index: <a href="../std/u64.md#std_u64">u64</a>) {
    <b>assert</b>!(bit_index &lt; bitvector.bit_field.<a href="../std/bit_vector.md#std_bit_vector_length">length</a>(), <a href="../std/bit_vector.md#std_bit_vector_EINDEX">EINDEX</a>);
    <b>let</b> x = &<b>mut</b> bitvector.bit_field[bit_index];
    *x = <b>true</b>;
}
</code></pre>



</details>

<a name="std_bit_vector_unset"></a>

## Function `unset`

Unset the bit at <code>bit_index</code> in the <code>bitvector</code> regardless of its previous state.


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_unset">unset</a>(bitvector: &<b>mut</b> <a href="../std/bit_vector.md#std_bit_vector_BitVector">std::bit_vector::BitVector</a>, bit_index: <a href="../std/u64.md#std_u64">u64</a>)
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_unset">unset</a>(bitvector: &<b>mut</b> <a href="../std/bit_vector.md#std_bit_vector_BitVector">BitVector</a>, bit_index: <a href="../std/u64.md#std_u64">u64</a>) {
    <b>assert</b>!(bit_index &lt; bitvector.bit_field.<a href="../std/bit_vector.md#std_bit_vector_length">length</a>(), <a href="../std/bit_vector.md#std_bit_vector_EINDEX">EINDEX</a>);
    <b>let</b> x = &<b>mut</b> bitvector.bit_field[bit_index];
    *x = <b>false</b>;
}
</code></pre>



</details>

<a name="std_bit_vector_shift_left"></a>

## Function `shift_left`

Shift the <code>bitvector</code> left by <code>amount</code>. If <code>amount</code> is greater than the
bitvector's length the bitvector will be zeroed out.


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_shift_left">shift_left</a>(bitvector: &<b>mut</b> <a href="../std/bit_vector.md#std_bit_vector_BitVector">std::bit_vector::BitVector</a>, amount: <a href="../std/u64.md#std_u64">u64</a>)
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_shift_left">shift_left</a>(bitvector: &<b>mut</b> <a href="../std/bit_vector.md#std_bit_vector_BitVector">BitVector</a>, amount: <a href="../std/u64.md#std_u64">u64</a>) {
    <b>if</b> (amount &gt;= bitvector.<a href="../std/bit_vector.md#std_bit_vector_length">length</a>) {
       <b>let</b> len = bitvector.bit_field.<a href="../std/bit_vector.md#std_bit_vector_length">length</a>();
       <b>let</b> <b>mut</b> i = 0;
       <b>while</b> (i &lt; len) {
           <b>let</b> elem = &<b>mut</b> bitvector.bit_field[i];
           *elem = <b>false</b>;
           i = i + 1;
       };
    } <b>else</b> {
        <b>let</b> <b>mut</b> i = amount;
        <b>while</b> (i &lt; bitvector.<a href="../std/bit_vector.md#std_bit_vector_length">length</a>) {
            <b>if</b> (bitvector.<a href="../std/bit_vector.md#std_bit_vector_is_index_set">is_index_set</a>(i)) bitvector.<a href="../std/bit_vector.md#std_bit_vector_set">set</a>(i - amount)
            <b>else</b> bitvector.<a href="../std/bit_vector.md#std_bit_vector_unset">unset</a>(i - amount);
            i = i + 1;
        };
        i = bitvector.<a href="../std/bit_vector.md#std_bit_vector_length">length</a> - amount;
        <b>while</b> (i &lt; bitvector.<a href="../std/bit_vector.md#std_bit_vector_length">length</a>) {
            <a href="../std/bit_vector.md#std_bit_vector_unset">unset</a>(bitvector, i);
            i = i + 1;
        };
    }
}
</code></pre>



</details>

<a name="std_bit_vector_is_index_set"></a>

## Function `is_index_set`

Return the value of the bit at <code>bit_index</code> in the <code>bitvector</code>. <code><b>true</b></code>
represents "1" and <code><b>false</b></code> represents a 0


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_is_index_set">is_index_set</a>(bitvector: &<a href="../std/bit_vector.md#std_bit_vector_BitVector">std::bit_vector::BitVector</a>, bit_index: <a href="../std/u64.md#std_u64">u64</a>): bool
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_is_index_set">is_index_set</a>(bitvector: &<a href="../std/bit_vector.md#std_bit_vector_BitVector">BitVector</a>, bit_index: <a href="../std/u64.md#std_u64">u64</a>): bool {
    <b>assert</b>!(bit_index &lt; bitvector.bit_field.<a href="../std/bit_vector.md#std_bit_vector_length">length</a>(), <a href="../std/bit_vector.md#std_bit_vector_EINDEX">EINDEX</a>);
    bitvector.bit_field[bit_index]
}
</code></pre>



</details>

<a name="std_bit_vector_length"></a>

## Function `length`

Return the length (number of usable bits) of this bitvector


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_length">length</a>(bitvector: &<a href="../std/bit_vector.md#std_bit_vector_BitVector">std::bit_vector::BitVector</a>): <a href="../std/u64.md#std_u64">u64</a>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_length">length</a>(bitvector: &<a href="../std/bit_vector.md#std_bit_vector_BitVector">BitVector</a>): <a href="../std/u64.md#std_u64">u64</a> {
    bitvector.bit_field.<a href="../std/bit_vector.md#std_bit_vector_length">length</a>()
}
</code></pre>



</details>

<a name="std_bit_vector_longest_set_sequence_starting_at"></a>

## Function `longest_set_sequence_starting_at`

Returns the length of the longest sequence of set bits starting at (and
including) <code>start_index</code> in the <code>bitvector</code>. If there is no such
sequence, then <code>0</code> is returned.


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_longest_set_sequence_starting_at">longest_set_sequence_starting_at</a>(bitvector: &<a href="../std/bit_vector.md#std_bit_vector_BitVector">std::bit_vector::BitVector</a>, start_index: <a href="../std/u64.md#std_u64">u64</a>): <a href="../std/u64.md#std_u64">u64</a>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_longest_set_sequence_starting_at">longest_set_sequence_starting_at</a>(bitvector: &<a href="../std/bit_vector.md#std_bit_vector_BitVector">BitVector</a>, start_index: <a href="../std/u64.md#std_u64">u64</a>): <a href="../std/u64.md#std_u64">u64</a> {
    <b>assert</b>!(start_index &lt; bitvector.<a href="../std/bit_vector.md#std_bit_vector_length">length</a>, <a href="../std/bit_vector.md#std_bit_vector_EINDEX">EINDEX</a>);
    <b>let</b> <b>mut</b> index = start_index;
    // Find the greatest index in the <a href="../std/vector.md#std_vector">vector</a> such that all indices less than it are <a href="../std/bit_vector.md#std_bit_vector_set">set</a>.
    <b>while</b> (index &lt; bitvector.<a href="../std/bit_vector.md#std_bit_vector_length">length</a>) {
        <b>if</b> (!bitvector.<a href="../std/bit_vector.md#std_bit_vector_is_index_set">is_index_set</a>(index)) <b>break</b>;
        index = index + 1;
    };
    index - start_index
}
</code></pre>



</details>


[//]: # ("File containing references which can be used from documentation")
