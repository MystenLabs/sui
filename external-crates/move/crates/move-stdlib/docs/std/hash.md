
<a name="std_hash"></a>

# Module `std::hash`

Module which defines SHA hashes for byte vectors.

The functions in this module are natively declared both in the Move runtime
as in the Move prover's prelude.


-  [Function `sha2_256`](#std_hash_sha2_256)
-  [Function `sha3_256`](#std_hash_sha3_256)


<pre><code></code></pre>



<a name="std_hash_sha2_256"></a>

## Function `sha2_256`



<pre><code><b>public</b> <b>fun</b> <a href="../std/hash.md#std_hash_sha2_256">sha2_256</a>(data: <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;): <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>native</b> <b>public</b> <b>fun</b> <a href="../std/hash.md#std_hash_sha2_256">sha2_256</a>(data: <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;): <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;;
</code></pre>



</details>

<a name="std_hash_sha3_256"></a>

## Function `sha3_256`



<pre><code><b>public</b> <b>fun</b> <a href="../std/hash.md#std_hash_sha3_256">sha3_256</a>(data: <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;): <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>native</b> <b>public</b> <b>fun</b> <a href="../std/hash.md#std_hash_sha3_256">sha3_256</a>(data: <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;): <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;;
</code></pre>



</details>


[//]: # ("File containing references which can be used from documentation")
