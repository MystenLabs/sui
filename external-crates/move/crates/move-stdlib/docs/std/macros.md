
<a name="std_macros"></a>

# Module `std::macros`

This module holds shared implementation of macros used in <code>std</code>




<pre><code></code></pre>


[//]: # ("File containing references which can be used from documentation")
