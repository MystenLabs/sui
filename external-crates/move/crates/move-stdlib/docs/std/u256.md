
<a name="std_u256"></a>

# Module `std::u256`



-  [Function `max`](#std_u256_max)
-  [Function `min`](#std_u256_min)
-  [Function `diff`](#std_u256_diff)
-  [Function `divide_and_round_up`](#std_u256_divide_and_round_up)
-  [Function `pow`](#std_u256_pow)


<pre><code></code></pre>



<a name="std_u256_max"></a>

## Function `max`

Return the larger of <code>x</code> and <code>y</code>


<pre><code><b>public</b> <b>fun</b> <a href="../std/u256.md#std_u256_max">max</a>(x: <a href="../std/u256.md#std_u256">u256</a>, y: <a href="../std/u256.md#std_u256">u256</a>): <a href="../std/u256.md#std_u256">u256</a>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/u256.md#std_u256_max">max</a>(x: <a href="../std/u256.md#std_u256">u256</a>, y: <a href="../std/u256.md#std_u256">u256</a>): <a href="../std/u256.md#std_u256">u256</a> {
    <a href="../std/macros.md#std_macros_num_max">std::macros::num_max</a>!(x, y)
}
</code></pre>



</details>

<a name="std_u256_min"></a>

## Function `min`

Return the smaller of <code>x</code> and <code>y</code>


<pre><code><b>public</b> <b>fun</b> <a href="../std/u256.md#std_u256_min">min</a>(x: <a href="../std/u256.md#std_u256">u256</a>, y: <a href="../std/u256.md#std_u256">u256</a>): <a href="../std/u256.md#std_u256">u256</a>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/u256.md#std_u256_min">min</a>(x: <a href="../std/u256.md#std_u256">u256</a>, y: <a href="../std/u256.md#std_u256">u256</a>): <a href="../std/u256.md#std_u256">u256</a> {
    <a href="../std/macros.md#std_macros_num_min">std::macros::num_min</a>!(x, y)
}
</code></pre>



</details>

<a name="std_u256_diff"></a>

## Function `diff`

Return the absolute value of x - y


<pre><code><b>public</b> <b>fun</b> <a href="../std/u256.md#std_u256_diff">diff</a>(x: <a href="../std/u256.md#std_u256">u256</a>, y: <a href="../std/u256.md#std_u256">u256</a>): <a href="../std/u256.md#std_u256">u256</a>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/u256.md#std_u256_diff">diff</a>(x: <a href="../std/u256.md#std_u256">u256</a>, y: <a href="../std/u256.md#std_u256">u256</a>): <a href="../std/u256.md#std_u256">u256</a> {
    <a href="../std/macros.md#std_macros_num_diff">std::macros::num_diff</a>!(x, y)
}
</code></pre>



</details>

<a name="std_u256_divide_and_round_up"></a>

## Function `divide_and_round_up`

Calculate x / y, but round up the result.


<pre><code><b>public</b> <b>fun</b> <a href="../std/u256.md#std_u256_divide_and_round_up">divide_and_round_up</a>(x: <a href="../std/u256.md#std_u256">u256</a>, y: <a href="../std/u256.md#std_u256">u256</a>): <a href="../std/u256.md#std_u256">u256</a>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/u256.md#std_u256_divide_and_round_up">divide_and_round_up</a>(x: <a href="../std/u256.md#std_u256">u256</a>, y: <a href="../std/u256.md#std_u256">u256</a>): <a href="../std/u256.md#std_u256">u256</a> {
    <a href="../std/macros.md#std_macros_num_divide_and_round_up">std::macros::num_divide_and_round_up</a>!(x, y)
}
</code></pre>



</details>

<a name="std_u256_pow"></a>

## Function `pow`

Return the value of a base raised to a power


<pre><code><b>public</b> <b>fun</b> <a href="../std/u256.md#std_u256_pow">pow</a>(base: <a href="../std/u256.md#std_u256">u256</a>, exponent: <a href="../std/u8.md#std_u8">u8</a>): <a href="../std/u256.md#std_u256">u256</a>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/u256.md#std_u256_pow">pow</a>(base: <a href="../std/u256.md#std_u256">u256</a>, exponent: <a href="../std/u8.md#std_u8">u8</a>): <a href="../std/u256.md#std_u256">u256</a> {
    <a href="../std/macros.md#std_macros_num_pow">std::macros::num_pow</a>!(base, exponent)
}
</code></pre>



</details>


[//]: # ("File containing references which can be used from documentation")
