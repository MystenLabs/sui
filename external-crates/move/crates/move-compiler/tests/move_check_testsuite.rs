// Copyright (c) The Diem Core Contributors
// Copyright (c) The Move Contributors
// SPDX-License-Identifier: Apache-2.0

use std::{
    collections::BTreeMap,
    fs,
    path::{Path, PathBuf},
};

use move_command_line_common::{
    env::read_bool_env_var,
    files::MOVE_EXTENSION,
    insta_assert,
    testing::{InstaOptions, OUT_EXT},
};
use move_compiler::{
    command_line::compiler::move_check_for_errors,
    diagnostics::warning_filters::WarningFiltersBuilder,
    diagnostics::*,
    editions::{Edition, Flavor},
    linters::{self, LintLevel},
    shared::{Flags, NumericalAddress, PackageConfig, PackagePaths},
    sui_mode, Compiler, PASS_PARSER,
};
use serde::{Deserialize, Serialize};

/// Shared flag to keep any temporary results of the test
const KEEP_TMP: &str = "KEEP";

const TEST_EXT: &str = "unit_test";
const UNUSED_EXT: &str = "unused";
const MIGRATION_EXT: &str = "migration";
const IDE_EXT: &str = "ide";

const LINTER_DIR: &str = "linter";
const SUI_MODE_DIR: &str = "sui_mode";
const MOVE_2024_DIR: &str = "move_2024";
const DEV_DIR: &str = "development";

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
struct TestInfo {
    flavor: Flavor,
    edition: Edition,
    lint: bool,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum TestKind {
    // Normal test
    Normal,
    // Tests unit test functionality
    Test,
    // Does not silence warnings for unused items
    Unused,
    // Tests edition migration
    Migration,
    // Tests additional generation for the IDE
    IDE,
}

impl TestKind {
    fn from_extension(path_extension: &std::ffi::OsStr) -> Self {
        match () {
            _ if path_extension == MOVE_EXTENSION => TestKind::Normal,
            _ if path_extension == TEST_EXT => TestKind::Test,
            _ if path_extension == UNUSED_EXT => TestKind::Unused,
            _ if path_extension == MIGRATION_EXT => TestKind::Migration,
            _ if path_extension == IDE_EXT => TestKind::IDE,
            _ => panic!("Unknown extension: {}", path_extension.to_string_lossy()),
        }
    }

    fn snap_suffix(&self) -> Option<&'static str> {
        match self {
            TestKind::Normal => None,
            TestKind::Test => Some(TEST_EXT),
            TestKind::Unused => Some(UNUSED_EXT),
            TestKind::Migration => Some(MIGRATION_EXT),
            TestKind::IDE => Some(IDE_EXT),
        }
    }
}

fn default_testing_addresses(flavor: Flavor) -> BTreeMap<String, NumericalAddress> {
    let mut mapping = vec![
        ("std", "0x1"),
        ("sui", "0x2"),
        ("M", "0x40"),
        ("A", "0x41"),
        ("B", "0x42"),
        ("K", "0x19"),
        ("a", "0x44"),
        ("b", "0x45"),
        ("k", "0x19"),
    ];
    if flavor == Flavor::Sui {
        mapping.extend([("sui", "0x2"), ("sui_system", "0x3")]);
    }
    mapping
        .into_iter()
        .map(|(name, addr)| (name.to_string(), NumericalAddress::parse_str(addr).unwrap()))
        .collect()
}

fn test_config(path: &Path) -> (TestKind, TestInfo, PackageConfig, Flags) {
    let test_kind = TestKind::from_extension(path.extension().unwrap());
    let path_contains = |s| path.components().any(|c| c.as_os_str() == s);
    let lint = path_contains(LINTER_DIR);
    let flavor = if path_contains(SUI_MODE_DIR) {
        Flavor::Sui
    } else {
        Flavor::default()
    };
    let move_2024_mode = path_contains(MOVE_2024_DIR);
    let dev_mode = path_contains(DEV_DIR);
    assert!(
        [move_2024_mode, dev_mode]
            .into_iter()
            .filter(|x| *x)
            .count()
            <= 1,
        "A test can have at most directory based edition"
    );
    let edition = if test_kind == TestKind::Migration {
        // migration mode overrides the edition
        Edition::E2024_MIGRATION
    } else if move_2024_mode {
        Edition::E2024_ALPHA
    } else if dev_mode {
        Edition::DEVELOPMENT
    } else {
        Edition::LEGACY
    };
    // config
    let mut config = PackageConfig {
        flavor,
        edition,
        is_dependency: false,
        warning_filter: WarningFiltersBuilder::new_for_source(),
    };
    // Unused and IDE do not have additional warning filters
    if !matches!(test_kind, TestKind::Unused | TestKind::IDE) {
        config
            .warning_filter
            .union(&WarningFiltersBuilder::unused_warnings_filter_for_test());
    }
    // test info
    let test_info = TestInfo {
        flavor,
        edition,
        lint,
    };
    // flags
    let flags = match test_kind {
        // no flags for normal tests
        TestKind::Normal => Flags::empty(),
        // we want to be able to see test/test_only elements in these modes
        TestKind::Test | TestKind::Unused | TestKind::Migration => Flags::testing(),
        // additional flags for IDE
        TestKind::IDE => Flags::testing().set_ide_test_mode(true).set_ide_mode(true),
    };
    (test_kind, test_info, config, flags)
}

fn out_path(path: &Path, test_name: &str, test_kind: Option<&str>) -> PathBuf {
    let n;
    let file_name = match test_kind {
        Some(c) => {
            n = format!("{test_name}@{c}");
            &n
        }
        None => test_name,
    };
    path.with_file_name(file_name).with_extension(OUT_EXT)
}

// Runs all tests under the test/testsuite directory.
pub fn run_test(path: &Path) -> datatest_stable::Result<()> {
    let (test_kind, test_info, package_config, flags) = test_config(path);
    let suffix = test_kind.snap_suffix();
    let migration_mode = package_config.edition == Edition::E2024_MIGRATION;
    let test_name = path.file_stem().unwrap().to_string_lossy();
    let test_name: &str = test_name.as_ref();
    let move_path = path.with_extension(MOVE_EXTENSION);
    let out_path = out_path(path, test_name, suffix);
    let flavor = package_config.flavor;
    let targets: Vec<String> = vec![move_path.to_str().unwrap().to_owned()];
    let named_address_map = default_testing_addresses(flavor);
    let deps = vec![PackagePaths {
        name: Some(("stdlib".into(), PackageConfig::default())),
        paths: move_stdlib::move_stdlib_files(),
        named_address_map: named_address_map.clone(),
    }];
    let target_name = if migration_mode {
        Some(("test".into(), package_config.clone()))
    } else {
        None
    };
    let targets = vec![PackagePaths {
        name: target_name,
        paths: targets,
        named_address_map,
    }];

    let flags = flags.set_sources_shadow_deps(true);
    let mut compiler = Compiler::from_package_paths(None, targets, deps)
        .unwrap()
        .set_flags(flags)
        .set_default_config(package_config);

    if flavor == Flavor::Sui {
        let (prefix, filters) = sui_mode::linters::known_filters();
        compiler = compiler.add_custom_known_filters(prefix, filters);
        if test_info.lint {
            compiler = compiler.add_visitors(sui_mode::linters::linter_visitors(LintLevel::All))
        }
    }
    let (prefix, filters) = linters::known_filters();
    compiler = compiler.add_custom_known_filters(prefix, filters);
    if test_info.lint {
        compiler = compiler.add_visitors(linters::linter_visitors(LintLevel::All))
    }

    let (files, comments_and_compiler_res) = compiler.run::<PASS_PARSER>()?;
    let diags = move_check_for_errors(comments_and_compiler_res);

    let has_diags = !diags.is_empty();
    let diag_buffer = if has_diags {
        if migration_mode {
            report_migration_to_buffer(&files, diags)
        } else {
            report_diagnostics_to_buffer(&files, diags, /* ansi_color */ false)
        }
    } else {
        vec![]
    };

    let save_diags = read_bool_env_var(KEEP_TMP);

    let rendered_diags = std::str::from_utf8(&diag_buffer)?;
    if save_diags {
        fs::write(out_path, &diag_buffer)?;
    }

    let mut options = InstaOptions::new();
    options.info(test_info);
    if let Some(suffix) = suffix {
        options.suffix(suffix);
    }
    options.name(test_name);
    insta_assert! {
        input_path: move_path,
        contents: rendered_diags,
        options: options,
    };
    Ok(())
}

datatest_stable::harness!(
    run_test,
    "tests/",
    r".*\.move$",
    run_test,
    "tests/",
    r".*\.unit_test$",
    run_test,
    "tests/",
    r".*\.unused$",
    run_test,
    "tests/",
    r".*\.migration$",
    run_test,
    "tests/",
    r".*\.ide$",
);
