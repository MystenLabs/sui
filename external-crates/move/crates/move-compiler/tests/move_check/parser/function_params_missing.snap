---
source: crates/move-compiler/tests/move_check_testsuite.rs
info:
  flavor: core
  edition: legacy
  lint: false
---
error[E01002]: unexpected token
  ┌─ tests/move_check/parser/function_params_missing.move:3:12
  │
3 │     fun fn { }
  │            ^
  │            │
  │            Unexpected '{'
  │            Expected '('
